<a href="https://colab.research.google.com/github/Surajk86808/ML-code/blob/main/Logistic_Regression_Imbalance_Dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
from collections import Counter
from sklearn.datasets import make_classification

In [5]:
# Generate sample classification data
X, y = make_classification(n_samples=1000, n_features=20, random_state=42)

# Display the shapes of the generated data
print("Shape of features (X):", X.shape)
print("Shape of target (y):", y.shape)

Shape of features (X): (1000, 20)
Shape of target (y): (1000,)


In [6]:
# Use Counter to count the occurrences of each class in the target variable
class_counts = Counter(y)

# Display the class counts
print("Class distribution:", class_counts)

Class distribution: Counter({np.int64(1): 500, np.int64(0): 500})


**Reasoning**:
Import the necessary function and split the data into training and testing sets as instructed.



In [7]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (800, 20)
Shape of X_test: (200, 20)
Shape of y_train: (800,)
Shape of y_test: (200,)


**Reasoning**:
Import LogisticRegression, instantiate the model, and train it using the training data.



In [8]:
from sklearn.linear_model import LogisticRegression

# Instantiate a LogisticRegression model with default parameters
model = LogisticRegression()

# Train the model using the training data
model.fit(X_train, y_train)

print("Logistic Regression model trained successfully.")

Logistic Regression model trained successfully.


**Reasoning**:
Import the necessary metrics functions from scikit-learn and then use the trained model to make predictions on the test set.



In [11]:
from sklearn.metrics import classification_report, confusion_matrix

# Make predictions on the test set
y_pred = model.predict(X_test)

**Reasoning**:
Generate and print the classification report and confusion matrix to evaluate the model's performance using the true and predicted labels.



In [12]:
# Generate and print the classification report
print("Classification Report:")
print(classification_report(y_test, y_pred))

# Generate and print the confusion matrix
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Classification Report:
              precision    recall  f1-score   support

           0       0.80      0.91      0.85        93
           1       0.91      0.80      0.86       107

    accuracy                           0.85       200
   macro avg       0.86      0.86      0.85       200
weighted avg       0.86      0.85      0.86       200


Confusion Matrix:
[[85  8]
 [21 86]]


## Summary:

### Data Analysis Key Findings

*   The dataset was split into training and testing sets with 80% of the data used for training and 20% for testing.
*   A Logistic Regression model was successfully trained on the training data.
*   The model achieved an accuracy of 0.85 on the test set.
*   The classification report shows good performance metrics (precision, recall, and F1-score) for both classes.

### Insights or Next Steps

*   The current model performs well on this balanced dataset.
*   While the dataset is balanced, the evaluation metrics chosen (precision, recall, F1-score, confusion matrix) are appropriate for assessing performance in cases of data imbalance, demonstrating a robust evaluation approach.
